### DISCUSSION TITLE
Need help.  Cava. Migration from automake to meson

### DISCUSSION THREAD

**Discussion by LukashonakV** [mesonbuild/meson#11562] [Q&A]
Dear all,
I'm trying to write meson wrapped file in order to be able to compile Cava via meson.
Here the [Cava upstream](https://github.com/karlstav/cava)
Original files for automake:

1. [configure.ac](https://github.com/karlstav/cava/blob/master/configure.ac)
2. [Makefile.am](https://github.com/karlstav/cava/blob/master/Makefile.am)

For this purposes I wrote : 

1. [meson_options.txt](https://github.com/mesonbuild/meson/files/11030193/meson_options.txt)
3. [meson.build.txt](https://github.com/mesonbuild/meson/files/11030200/meson.build.txt)

It compiles without issues....  
But the result is totally unexpected for me. On the left: cava is built by automake. On the right: cava is build by meson.
I've totally stuck and have no idea.. what is the reason. Can someone help me please ?
I've already tried many variants: turn off unnecessary options... even tried to cut of automake configuration... but despite everything automake does compilation Ok.. but meson doesn't.
![ps_2023-03-21-18_47_50](https://user-images.githubusercontent.com/23121044/226663932-2ce45994-cbd7-48cf-b0c1-aaf66bcc63b5.png)
![ps_2023-03-21-18_47_53](https://user-images.githubusercontent.com/23121044/226663981-fce8b5e6-42b5-4ab2-a9c7-456a60c9610f.png)
![ps_2023-03-21-18_47_53](https://user-images.githubusercontent.com/23121044/226664033-8b901bfb-add8-4f5a-9e40-737c8beab4ed.png)

**Comment by xclaesse** [comment]
You should compare the command line run by ninja and make to see if there is a different. Also if there is aconfig.h generated, compare what they define.

**Reply by LukashonakV** [reply]
Hi @xclaesse , actually I'm  stuck
aconfig.h is absent
These are autotool logs:
1. [build.log](https://github.com/mesonbuild/meson/files/11051346/build.log)
2. [config.log](https://github.com/mesonbuild/meson/files/11051348/config.log)

These are meson logs:
1. [meson.log](https://github.com/mesonbuild/meson/files/11051358/meson.log)
2. [build.ninja.txt](https://github.com/mesonbuild/meson/files/11051364/build.ninja.txt)
3. [meson-log.txt](https://github.com/mesonbuild/meson/files/11051527/meson-log.txt)
----
I did some additional fixies into meson.build in order to make build command line pretty simple to autotools
[meson.build.txt](https://github.com/mesonbuild/meson/files/11051390/meson.build.txt)
But again not any luck.

For example:
autotool
```log
gcc -DPACKAGE_NAME=\"cava\" -DPACKAGE_TARNAME=\"cava\" -DPACKAGE_VERSION=\"0.8.3-28-g2d212b5-dirty\" -DPACKAGE_STRING=\"cava\ 0.8.3-28-g2d212b5-dirty\" -DPACKAGE_BUGREPORT=\"karl@stavestrand.no\" -DPACKAGE_URL=\"\" -DPACKAGE=\"cava\" -DVERSION=\"0.8.3-28-g2d212b5-dirty\" -DHAVE_STDIO_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_STRINGS_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_UNISTD_H=1 -DSTDC_HEADERS=1 -DHAVE_DLFCN_H=1 -DLT_OBJDIR=\".libs/\" -I.  -DPACKAGE=\"cava\" -DVERSION=\"0.8.3-28-g2d212b5-dirty\" -DFONTDIR="" -DNDEBUG -DPULSE -I/usr/include/iniparser4  -g -O2 -MT cava-cava.o -MD -MP -MF .deps/cava-cava.Tpo -c -o cava-cava.o `test -f 'cava.c' || echo './'`cava.c
cava.c: In function ‘main’:
cava.c:306:21: warning: ignoring return value of ‘system’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
  306 |                     system("setfont " FONTDIR "/cava.psf  >/dev/null 2>&1");
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cava.c:309:21: warning: ignoring return value of ‘system’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
  309 |                     system("setfont cava.psf  >/dev/null 2>&1");
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cava.c:311:17: warning: ignoring return value of ‘system’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
  311 |                 system("setterm -blank 0");
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~
mv -f .deps/cava-cava.Tpo .deps/cava-cava.Po
```
Looks quite the same as in meson
```log
build cava.p/cava.c.o: c_COMPILER ../cava.c
 DEPFILE = cava.p/cava.c.o.d
 DEPFILE_UNQUOTED = cava.p/cava.c.o.d
 ARGS = -Icava.p -I. -I.. -I../input -I../output -I/usr/include/iniparser4 -I/usr/include/ncursesw -I/usr/include/SDL2 -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wextra -Wpedantic -std=c99 -O0 -g '-DPACKAGE="cava"' '-DVERSION="0.8.3"' '-DFONTDIR="/usr/share/consolefonts"' -D_POSIX_SOURCE -D_POSIX_C_SOURCE=200809L -D_XOPEN_SOURCE_EXTENDED -DHAVE_STDIO_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_STRINGS_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_UNISTD_H=1 -DSTDC_HEADERS=1 -DHAVE_DLFCN_H=1 '-DLT_OBJDIR=".libs/"' -DHAVE_ALLOCA_H -DALSA -DPULSE -DNCURSES -DSDL -DSDL_GLSL -Dc_std=c99 -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=600 -D_REENTRANT -pthread -std=c99 -Wall -Wextra -Wno-unused-result -Wno-maybe-uninitialized -Wno-vla-parameter -MD -MP -MF -MT
```

The only real difference I found is a Linker command line.
autotool
```log
/bin/sh ./libtool  --tag=CC   --mode=link gcc  -g -O2   -o cava cava-cava.o cava-cavacore.o cava-config.o input/cava-common.o input/cava-fifo.o input/cava-shmem.o output/cava-terminal_noncurses.o output/cava-raw.o output/cava-noritake.o   input/cava-pulse.o      -lpulse-simple -lpulse -lm -lfftw3 -liniparser4
libtool: link: gcc -g -O2 -o cava cava-cava.o cava-cavacore.o cava-config.o input/cava-common.o input/cava-fifo.o input/cava-shmem.o output/cava-terminal_noncurses.o output/cava-raw.o output/cava-noritake.o input/cava-pulse.o  -lpulse-simple -lpulse -lm -lfftw3 -liniparser4
```

Meson
```log
build cava: c_LINKER cava.p/cava.c.o cava.p/cavacore.c.o cava.p/config.c.o cava.p/input_common.c.o cava.p/input_fifo.c.o cava.p/input_shmem.c.o cava.p/output_terminal_noncurses.c.o cava.p/output_raw.c.o cava.p/output_noritake.c.o cava.p/input_alsa.c.o cava.p/input_pulse.c.o cava.p/output_terminal_ncurses.c.o cava.p/output_sdl_cava.c.o cava.p/output_sdl_glsl.c.o | /usr/lib64/libGL.so /usr/lib64/libSDL2.so /usr/lib64/libasound.so.2.0.0 /usr/lib64/libfftw3.so /usr/lib64/libiniparser4.so.1.0.0 /usr/lib64/libm.a /usr/lib64/libm.so /usr/lib64/libncursesw.so /usr/lib64/libpthread.a /usr/lib64/libpulse-simple.so.0.1.1 /usr/lib64/libpulse.so /usr/lib64/libtinfow.so
 LINK_ARGS = -Wl,--as-needed -Wl,--no-undefined -Wl,--start-group -lpthread -pthread -liniparser4 /usr/lib64/libfftw3.so -lm -lasound /usr/lib64/libpulse.so -lpulse-simple /usr/lib64/libncursesw.so /usr/lib64/libtinfow.so /usr/lib64/libSDL2.so /usr/lib64/libGL.so -Wl,--end-group
```

Can you please help me... Actually I have no idea here.. even still digging into it

**Reply by xclaesse** [reply]
There is a big difference: `-O0`, you can add `buildtype=debugoptimized` in your default_options.

**Reply by LukashonakV** [reply]
Yeah , I've already had done this... but still not any luck.
Now results the same. 
The same even I turned off c_std=c99

But interesting.... 
1. Shows 02 [build.ninja.txt](https://github.com/mesonbuild/meson/files/11052386/build.ninja.txt)
2. Shows -O0 [meson-log.txt](https://github.com/mesonbuild/meson/files/11052393/meson-log.txt) - looks like it's for the find_library or dependencies meson feature purposes only

![ps_2023-03-23-18_03_33](https://user-images.githubusercontent.com/23121044/227246050-67179ad4-ada5-4d98-9cb8-52d17c4cefd0.png)

**Reply by LukashonakV** [reply]
Hallelujah, for some reason 
```meson1
add_project_arguments('-DNDEBUG', language: 'c')
```
solved the issue
Thank you @xclaesse
